hail/src/main/scala/is/hail/expr/ir/streams/EmitStream.scala

```diff
@@ -1452,6 +1455,192 @@ object EmitStream {
           SStreamValue(producer)
         }
 
+      case x@StreamLeftIntervalJoin(left, right, lKeyField, rIntrvlName, lName, rName, body) =>
+        produce(left, cb).flatMap(cb) { case lStream: SStreamValue =>
+          produce(right, cb).map(cb) { case rStream: SStreamValue =>
+
+            // map over the keyStream
+            val lProd = lStream.getProducer(mb)
+            val rProd = rStream.getProducer(mb)
+
+            val rElemSTy = SBaseStructPointer(rProd.element.st.storageType().asInstanceOf[PBaseStruct])
+            val intrvlSTy = rElemSTy.fieldTypes(rElemSTy.fieldIdx(rIntrvlName))
+
+            val loadInterval: EmitMethodBuilder[_] =
+              mb.ecb.defineEmitMethod(genName("m", "LoadInterval"),
+                FastSeq(rElemSTy.paramType),
+                intrvlSTy.paramType
+              ) { mb =>
+                mb.emitSCode { cb =>
+                  mb.getSCodeParam(1)
+                    .asBaseStruct
+                    .loadField(cb, rIntrvlName)
+                    .get(cb)
+                }
+              }
+
+            val q: StagedMinHeap =
+              StagedMinHeap(mb.emodb, rElemSTy) { heapClassBuilder =>
+                new StagedMinHeap.StagedComparator {
+                  val parent: ThisFieldRef[_] =
+                    heapClassBuilder.genFieldThisRef("ParentRef")(mb.cb.ti)
+
+                  override def init(cb: EmitCodeBuilder, enclosingRef: Value[AnyRef]): Unit =
+                    cb.assignAny(parent, Code.checkcast(enclosingRef)(mb.cb.ti))
+
+                  override def apply(cb: EmitCodeBuilder, a: SValue, b: SValue): Value[Int] = {
+                    val l = cb.invokeSCode(loadInterval, parent, a).asInterval
+                    val r = cb.invokeSCode(loadInterval, parent, b).asInterval
+                    IntervalFunctions.intervalEndpointCompare(cb,
+                      l.loadEnd(cb).get(cb), l.includesEnd,
+                      r.loadEnd(cb).get(cb), r.includesEnd
+                    )
+                  }
+                }
+              }(mb.ecb)
```

Comment:
I'm surprised by the heaviness of the comparison function. Did you try having the StagedMinHeap take a function `(EmitCodeBuilder, SValue, SValue) => Value[Int]` (which the min heap could either inline at the call site or wrap in a method as desired)?

Reply:
Yes. The reason for this is to share the `loadInterval` code. Now it's inlined, there's no need.